junixsocket-common/src/main/java/org/newsclub/net/unix/AFSelector.java

```diff
@@ -110,47 +111,51 @@ public int select() throws IOException {
 
   @SuppressWarnings("PMD.CognitiveComplexity")
   private int select0(int timeout) throws IOException {
-    PollFd pfd;
     synchronized (this) {
       if (!isOpen()) {
         throw new ClosedSelectorException();
       }
-      pfd = pollFd = initPollFd(pollFd);
-      performRemove();
-      selectedKeysSet.clear();
-    }
-    int num;
-    try {
-      begin();
-      num = NativeUnixSocket.poll(pfd, timeout);
-    } finally {
-      end();
-    }
-    synchronized (this) {
-      selectedKeysSet.clear();
-      pfd = pollFd;
-      if (pfd != null) {
-        AFSelectionKey[] keys = pfd.keys;
-        if (keys != null) {
-          for (AFSelectionKey key : keys) {
-            if (key != null && key.hasOpInvalid()) {
-              SelectableChannel ch = key.channel();
-              if (ch != null && ch.isOpen()) {
-                ch.close();
+
+      synchronized (publicSelectedKeys) {
+        PollFd pfd = pollFd = initPollFd(pollFd);
+        processDeregisterQueue();
+        publicSelectedKeys.clear();
+
+        int num;
+        try {
+          begin();
+          num = NativeUnixSocket.poll(pfd, timeout);
+        } finally {
+          end();
+        }
+
+        publicSelectedKeys.clear();
+        pfd = pollFd;
+        if (pfd != null) {
+          AFSelectionKey[] keys = pfd.keys;
+          if (keys != null) {
+            for (AFSelectionKey key : keys) {
+              if (key != null && key.hasOpInvalid()) {
+                SelectableChannel ch = key.channel();
+                if (ch != null && ch.isOpen()) {
+                  ch.close();
+                }
               }
             }
           }
         }
+        if (num > 0) {
+          consumeAllBytesAfterPoll();
+          setOpsReady(pfd); // updates keysSelected and numKeysSelected
+        }
+        return publicSelectedKeys.size();
       }
-      if (num > 0) {
-        consumeAllBytesAfterPoll();
-        setOpsReady(pfd); // updates keysSelected and numKeysSelected
-      }
-      return selectedKeysSet.size();
     }
   }
 
-  private synchronized void consumeAllBytesAfterPoll() throws IOException {
+  private void consumeAllBytesAfterPoll() throws IOException {
+    assert Thread.holdsLock(this);
```

Comment:
Shouldn't this be assert Thread.holdsLock(publicSelectedKeys)) ?

Reply:
This case is interesting. It doesn't actually need to lock on `publicSelectedKeys` for correctness, but then we would have to lock, drop, then immediately lock on `publicSelectedKeys` again if we didn't want to consume bytes under the lock. I think we can reorder some of this to reduce what is under the lock.